hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/RealtimeUnmergedRecordReader.java

```diff
@@ -76,42 +72,50 @@ public RealtimeUnmergedRecordReader(RealtimeSplit split, JobConf job,
         Option.empty(), Function.identity(), new DefaultSizeEstimator<>(), Functions.noop());
     // Consumer of this record reader
-    this.iterator = this.executor.getQueue().iterator();
-    this.logRecordScanner = HoodieUnMergedLogRecordScanner.newBuilder()
-        .withFileSystem(FSUtils.getFs(split.getPath().toString(), this.jobConf))
-        .withBasePath(split.getBasePath())
-        .withLogFilePaths(split.getDeltaLogPaths())
-        .withReaderSchema(getReaderSchema())
-        .withLatestInstantTime(split.getMaxCommitTime())
-        .withReadBlocksLazily(Boolean.parseBoolean(this.jobConf.get(HoodieRealtimeConfig.COMPACTION_LAZY_BLOCK_READ_ENABLED_PROP, HoodieRealtimeConfig.DEFAULT_COMPACTION_LAZY_BLOCK_READ_ENABLED)))
-        .withReverseReader(false)
-        .withBufferSize(this.jobConf.getInt(HoodieRealtimeConfig.MAX_DFS_STREAM_BUFFER_SIZE_PROP, HoodieRealtimeConfig.DEFAULT_MAX_DFS_STREAM_BUFFER_SIZE))
-        .withLogRecordScannerCallback(record -> {
-          // convert Hoodie log record to Hadoop AvroWritable and buffer
-          GenericRecord rec = (GenericRecord) record.getData().getInsertValue(getReaderSchema(), payloadProps).get();
-          ArrayWritable aWritable = (ArrayWritable) HoodieRealtimeRecordReaderUtils.avroToArrayWritable(rec, getHiveSchema());
-          this.executor.getQueue().insertRecord(aWritable);
-        })
-        .build();
+    this.iterator = this.executor.getRecordIterator();
+
     // Start reading and buffering
-    this.executor.startProducers();
+    this.executor.startProducingAsync();
   }
 
   /**
    * Setup log and parquet reading in parallel. Both write to central buffer.
    */
-  private List<HoodieProducer<ArrayWritable>> getParallelProducers() {
-    List<HoodieProducer<ArrayWritable>> producers = new ArrayList<>();
-    producers.add(new FunctionBasedQueueProducer<>(buffer -> {
-      logRecordScanner.scan();
-      return null;
-    }));
-    producers.add(new IteratorBasedQueueProducer<>(parquetRecordsIterator));
-    return producers;
+  private List<HoodieProducer<ArrayWritable>> getParallelProducers(
+      HoodieUnMergedLogRecordScanner.Builder scannerBuilder
+  ) {
+    return Arrays.asList(
+        new FunctionBasedQueueProducer<>(queue -> {
+          HoodieUnMergedLogRecordScanner scanner =
+              scannerBuilder.withLogRecordScannerCallback(record -> {
+                    // convert Hoodie log record to Hadoop AvroWritable and buffer
+                    GenericRecord rec = (GenericRecord) record.getData().getInsertValue(getReaderSchema(), payloadProps).get();
+                    ArrayWritable aWritable = (ArrayWritable) HoodieRealtimeRecordReaderUtils.avroToArrayWritable(rec, getHiveSchema());
+                    queue.insertRecord(aWritable);
+                  })
+                  .build();
+          // Scan all the delta-log files, filling in the queue
+          scanner.scan();
```

Comment:
maybe assign scanner progress here, which can be used in L157.

Reply:
Will take this up in https://github.com/apache/hudi/pull/7245